src/runtime/runtime_rpi3.go

```diff
@@ -0,0 +1,140 @@
+// +build rpi3
+
+package runtime
+
+import (
+	"machine"
+	"runtime/volatile"
+	"unsafe"
+	"device/arm"
+)
+
+const tickMicros = int64(1)
+
+type timeUnit int64
+
+//go:export sleepticks sleepticks
+func sleepTicks(n timeUnit) {
+	WaitMuSec(uint32(n))
+}
+
+func ticks() timeUnit {
+	return timeUnit(SysTimer())
+}
+
+func preinit() {
+	//XXX because interp.Unsupported would be thrown if we tried to get this to happen at initializer time
+	machine.Aux = (*machine.AuxPeripheralsRegisterMap)(unsafe.Pointer(machine.MemoryMappedIO + 0x00215000))
+	machine.GPIO = (*machine.GPIORegisterMap)(unsafe.Pointer(machine.MemoryMappedIO + 0x00200000))
+	machine.SysTimer = (*machine.SysTimerRegisterMap)(unsafe.Pointer(machine.MemoryMappedIO + 0x3000))
+
+
+
+	// Initialize .bss: zero-initialized global variables.
+	//ptr := unsafe.Pointer(&_sbss)
+	//for ptr != unsafe.Pointer(&_ebss) {
+	//	*(*uint32)(ptr) = 0
+	//	ptr = unsafe.Pointer(uintptr(ptr) + 4)
+	//}
+
+	machine.UARTInit()
+}
+
+// main on bare metal (bootloader uses this path)
+//go:export main
+func main() {
+	preinit()
+	primaryMain()
+}
+
+func primaryMain() {
+	initAll()
+	callMain()
+	machine.Abort()
+}
+
+//go:extern
+func putchar(c byte) {
+	//MiniUARTSend(c) if you prefer the mini uart
+	machine.UARTPutc(c)
+}
+
+//
+// Wait MuSec waits for at least n musecs based on the system timer.
+//
+func WaitMuSec(n uint32) {
+	var f, t, r uint64
+	arm.AsmFull(`mrs x28, cntfrq_el0
+		str x28,{f}
+		mrs x27, cntpct_el0
+		str x27,{t}`, map[string]interface{}{"f": &f, "t": &t})
+	//expires at t
+	t += ((f / 1000) * uint64(n)) / 1000
+	for r < t {
+		arm.AsmFull(`mrs x27, cntpct_el0
+			str x27,{r}`, map[string]interface{}{"r": &r})
+ 	}
+}
+
+//
+// SysTimer gets the 64 bit timer's value.
+//
+func SysTimer() uint64 {
```

Comment:
Does this need to be exported?

If this needs to be exported, maybe put it in `machine`.

Reply:
Done.